syllabus/4-Smart_Contracts/6-Wasm_ink!_slides.md

```diff
@@ -636,48 +714,61 @@ Notes:
 
 ---
 
-## Contracts: Defining shared behaviour
+## Importance of events
 
-<div style="font-size: 0.5em;">
+* **Events** are information that is emitted by transactions taking place on a blockchain.
+* **Events** are way for a smart contracts to communicate that a specific action or state change has occurred.
 
-```rust [1-14|17,22]
-#[ink::trait_definition]
-pub trait PSP22 {
-    #[ink(message)]
-    fn total_supply(&self) -> Balance;
+Notes:
+- we talked about the technical aspects of events but what are they exactly and what are they usefull for
+- what happens when a tx is submitted?
+  - it goes to mempool where it is picked up by a block producer and eventually included in the blockchain.
+- so transactions do not immediately return a value. Enter events
 
-    #[ink(message)]
-    fn balance_of(&self, owner: AccountId) -> Balance;
+---
 
-    #[ink(message)]
-    fn approve(&mut self, spender: AccountId, amount: Balance) -> Result<(), PSP22Error>;
+## Importance of events
 
-    #[ink(message)]
-    fn transfer(&mut self, to: AccountId, value: Balance, data: Vec<u8>) -> Result<(), PSP22Error>;
-    ...
+<img rounded style="width: 300px;" src="./img/ink/transacting.svg" />
 
-impl SimpleDex {
-    use psp22_trait::{PSP22Error, PSP22};
+* The most important use of events is to pass along return values from contracts to a dapp's UI.
+* How to access historical data?
 
-    /// Returns balance of a PSP22 token for an account
-    fn balance_of(&self, token: AccountId, account: AccountId) -> Balance {
-        let psp22: ink::contract_ref!(PSP22) = token.into();
-        psp22.balance_of(account)
-    }
-    ...
-```
+Notes:
+- ppl are transacting among each other e.g. an PSP22 token
+- head of the contract state stores just their cumulative balance
+- what if you want to display a history of transactions?
+ - e.g. DEX may want to show to a user all the deposits they have made
+ - replaying txs is slow and expensive
+ - enter indexers
 
-</div>
+---
 
-- Trait Definition: `#[ink::trait_definition]`.
-- Sharing the trait definition to do a cross-contract call.
+## Event indexers
+
+* blockchain = append-only database
+* events = logs
+* indexers = log processors
+
+An **indexer** is a process that listens to the events, processes them and stores normalized data in a persistent, queryable storage.
+Typically some sort of API is then server over that data, that can be easily queried by the frontend applications.
+
+**Examples:**
+
+* [Subsquid: https://subsquid.io/](https://subsquid.io/)
+* [TheGraph: https://thegraph.com/](https://thegraph.com/)
+
+<!-- TODO A cheaper form of storage -->
```

Comment:
Is this TODO something that will be added for this cohort? As part of this lesson?
If so - then I think it is preferable to finalize the presentation in this PR instead of leaving TODOs

Reply:
done, separated into a slide